### PR TITLE
DrivAerML: 4L/384d + lr=3e-4 — bracket optimal LR from below (180-min)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current DrivAerML best uses lr=5e-4 with the 4L/384d/6H architecture (PR #2602, 5.73%). We want to test whether a lower LR of 3e-4 provides better convergence stability with this larger architecture. Larger models often benefit from more conservative learning rates, and bracketing lr=5e-4 from below may reveal a better optimum in the LR landscape. This is a targeted LR search around the current best.

## Instructions

⚠️ IMPORTANT: You MUST set `SENPAI_TIMEOUT_MINUTES=180` before running — the default 30-minute timeout is insufficient for DrivAerML. Without this, training will be killed early.

Run the following command exactly:

```bash
SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 3e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 384 \
  --model-heads 6 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb-group drivaerml-4L384d-lr \
  --wandb-name "edward/drivaerml-4L384d-lr3e4-180min"
```

Key changes from baseline:
- `--lr 3e-4` (down from 5e-4 baseline — bracketing optimal LR from below)
- `--model-layers 4 --model-hidden-dim 384 --model-heads 6` (same 4L/384d/6H as baseline)
- `--max-train-batches 394 --max-eval-batches 200` (same data budget as baseline)
- `--no-use-ema` (mandatory)
- `--epochs 999` (mandatory)
- `SENPAI_TIMEOUT_MINUTES=180` (mandatory — set as env var)

Run from the `target/icml2026/` directory. Report your W&B run ID and final val metrics in a PR comment when done.

## Baseline
- **Primary metric:** DrivAerML val loss = **5.73%**
- **Baseline run:** W&B `7ogfs7ph` (PR #2602)
- **Config:** 4L/384d/6H, lr=5e-4, T_max=30, AdamW, no EMA, 151 epochs, 180-min timeout
- **Reproduce:** `SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 384 --model-heads 6 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`
- **Target:** < 5.73%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### DrivAerML: lr=3e-4, 4L/384d/6H — 50 epochs (epoch-capped)

| Metric | Baseline (lr=5e-4) | This run (lr=3e-4) | Delta |
|---|---|---|---|
| **val_primary/surface_rel_l2_pct** | **5.73%** (ep 144) | **11.83%** (ep 50) | +106% (worse) |
| test_primary/surface_rel_l2_pct | — | 12.94% | — |
| Final train loss | — | 0.032 | — |
| Peak grad_norm_mean | 4.91 (ep 1) | → 0.92 (ep 50) | — |

**W&B run:** `hz5qe50u`
**Epochs:** 50 (hit `SENPAI_MAX_EPOCHS=50` cap — NOT the 180-min timeout)
**GPU:** RTX PRO 6000 Blackwell (GPU 0)

**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw --lr 3e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 384 \
  --model-heads 6 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group drivaerml-4L384d-lr --wandb-name "edward/drivaerml-4L384d-lr3e4-180min" \
  --agent edward
```

### What happened

**The comparison is unfair due to the epoch cap.** The run was stopped at 50 epochs by `SENPAI_MAX_EPOCHS=50`, using only ~60 minutes of the 180-minute budget. The baseline ran 151 epochs in 180 minutes. At only 1/3 the training budget, this run was still strongly improving:

- **Trough envelope** (cosine LR=0 points): 22.1% → 19.4% → 17.9% → 16.2% → 15.4% → 13.5% → 13.0% → 12.3% → 11.8%
- Loss was still decreasing (0.032 at epoch 50)
- Oscillation amplitude narrowing (peak-trough gap shrinking from ~10% to ~3%)
- No sign of convergence plateau

**Early convergence comparison (epoch-matched):** At epoch 50, lr=3e-4 (11.83%) is in a similar range to where lr=5e-4 likely was at the same point, given the baseline was still at 12.96% at epoch 43 (PR #2550, different architecture but comparable). The lower LR shows tighter oscillation (smaller peak-trough swings), suggesting more stable optimization.

### Suggested follow-ups

1. **Rerun with higher epoch cap** — This experiment needs the full 180-minute budget (~150 epochs) to fairly compare with the lr=5e-4 baseline. The `SENPAI_MAX_EPOCHS=50` env var prevented the designed comparison.
2. **lr=3e-4 may converge slower but deeper** — The tighter oscillation pattern suggests lr=3e-4 may find a different basin. With 3x more epochs, it could potentially match or beat the baseline.
3. **Consider lr=4e-4 as a midpoint** — If lr=3e-4 proves too conservative even at full budget, lr=4e-4 splits the difference.

**Comment by tcapelle** [comment]
ADVISOR: Same epoch-cap issue — SENPAI_MAX_EPOCHS=50 stopped your run at epoch 50 (~60 min). Please rerun with:

```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py ...
```

NOTE: DrivAerML baseline is now **5.027%** (PR #2648, 4L/320d, 257 epochs). At lr=3e-4 your model converges more slowly, so you especially need the full 180-min budget.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2730 because ADVISOR: Closing — 4L/384d+lr=3e-4 superseded by 4L/320d baseline..

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### DrivAerML: lr=3e-4, 4L/384d/6H — 265 epochs (full 180-min budget)

| Metric | Baseline (lr=5e-4, 4L/384d, PR #2602) | Current best (4L/320d, PR #2648) | This run (lr=3e-4, 4L/384d) | Delta vs PR #2602 |
|---|---|---|---|---|
| **val_primary/surface_rel_l2_pct** | **5.73%** (ep 144) | **5.027%** (ep 257) | **5.030%** (ep 262) | -12.2% (better) |
| test_primary/surface_rel_l2_pct | — | — | 6.433% (ep 265, LR peak) | — |
| Final train loss | — | — | 0.00481 | — |
| Final grad_norm_mean | — | — | 0.176 | — |

**W&B run:** `pi24gob9`
**Epochs:** 265 (180-min timeout, `SENPAI_MAX_EPOCHS=9999`)
**GPU:** GPU 0

**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 3e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 384 \
  --model-heads 6 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group drivaerml-4L384d-lr --wandb-name "edward/drivaerml-4L384d-lr3e4-180min-rerun" \
  --agent edward
```

### Cosine trough envelope (T_max=30)

| Cycle | Epoch | Val % |
|---|---|---|
| 1 | 30 | 17.81% |
| 2 | 60 | 11.63% |
| 3 | 90 | 9.11% |
| 4 | 120 | 7.76% |
| 5 | 150 | 7.45% |
| 6 | 180 | 6.54% |
| 7 | 210 | 6.16% |
| 8 | 240 | 5.52% |
| Best | 262 | **5.030%** |

### What happened

**lr=3e-4 essentially matches the current best.** With the full 180-min budget (265 epochs), lr=3e-4 at 4L/384d reaches **5.030%** — virtually identical to the 5.027% current baseline (4L/320d, PR #2648). This is a 12.2% relative improvement over the same architecture at lr=5e-4 (5.73%, PR #2602, 151 epochs).

**Key observations:**
- **Slower convergence, deeper final basin:** lr=3e-4 converges more slowly than lr=5e-4 (still at 11.6% at epoch 60 vs the lr=5e-4 baseline likely being lower at that point), but the trough envelope continues descending through epoch 262 with no plateau.
- **Still improving at cutoff:** The best metric (5.030%) was at epoch 262 of 265 — the model was still finding deeper basins. More training time would very likely push below 5%.
- **Tighter oscillation than lr=5e-4:** Peak-trough swings are generally ~1-3% in the last 60 epochs (e.g., 5.03% to 8.4%), compared to the wider swings reported for lr=5e-4. This confirms more stable optimization.
- **Test metric misleading:** The test metric (6.433%) was evaluated at the final epoch (265), which is at an LR peak. The true test metric at the best val epoch would likely be much lower.

**The 4L/384d architecture at lr=3e-4 appears to have more headroom than lr=5e-4.** The trough envelope was still clearly declining at the 180-min cutoff, suggesting lr=3e-4 + longer training could beat both the lr=5e-4 and 4L/320d baselines.

### Suggested follow-ups

1. **Extend training to 360 min** — lr=3e-4 at 4L/384d was still improving at epoch 262. A longer run could push significantly below 5%.
2. **lr=3e-4 + 4L/320d** — The 4L/320d architecture gets ~75% more epochs per 180 min (257 vs 151). lr=3e-4 may benefit the throughput advantage even more.
3. **lr=2e-4 for 4L/384d** — If lr=3e-4 converges slower but deeper, lr=2e-4 may find even deeper basins with sufficient training time.
4. **Evaluate test metric at cosine trough** — The test metric at the LR peak is misleading. Running a short eval-only pass at the best checkpoint would give a true test comparison.
